python/dpu_utils/ptutils/componenttrainer.py

```diff
@@ -0,0 +1,172 @@
+            self.__model.train()
+
+            data_iter = shuffled_iterator(training_tensors())
+            sum_epoch_loss = 0.0
+            running_avg_loss = 0.0
+            num_minibatches = 0
+            num_samples = 0
+            have_more_data = True
+            start_time = time.time()
+            self.__model.reset_metrics()
+            with tqdm(desc='Training', disable=not show_progress_bar, leave=False) as progress_bar:
+                while have_more_data:
+                    mb_data, have_more_data, num_elements = self.__model.create_minibatch(data_iter,
+                                                                                          max_num_items=self.__minibatch_size)
+                    if not have_more_data or num_elements == 0:
+                        break  # Do not consider half-full minibatches
+                    optimizer.zero_grad()
+                    mb_loss = self.__model.forward(**mb_data)
+                    mb_loss.backward()
+
+                    optimizer.step()
+                    num_minibatches += 1
+                    num_samples += num_elements
+                    sum_epoch_loss += float(mb_loss.cpu())
+                    if num_minibatches == 1:  # First minibatch
+                        running_avg_loss = float(mb_loss.cpu())
+                    else:
+                        running_avg_loss = exponential_running_average_factor * running_avg_loss + (
+                                    1 - exponential_running_average_factor) * float(mb_loss.cpu())
+                    progress_bar.update()
+                    progress_bar.set_postfix(Loss=f'{running_avg_loss:.2f}')
+
+            elapsed_time = time.time() - start_time  # type: float
+            self.LOGGER.info('Training complete in %.1fsec [%.2f samples/sec]', elapsed_time,
+                             (num_samples / elapsed_time))
+            self.LOGGER.info('Epoch %i: Avg Train Loss %.2f', epoch + 1, sum_epoch_loss / num_minibatches)
+            train_metrics = self.__model.report_metrics()
+            if len(train_metrics) > 0:
+                self.LOGGER.info('Training Metrics: %s', json.dumps(train_metrics, indent=2))
+
+            # Now do validation!
+            self.__model.eval()
+            data_iter = validation_tensors()
+            sum_epoch_loss = 0
+            num_minibatches = 0
+            num_samples = 0
+            have_more_data = True
+            start_time = time.time()
+            self.__model.reset_metrics()
+            with tqdm(desc='Validation', disable=not show_progress_bar, leave=False) as progress_bar, torch.no_grad():
+                while have_more_data:
+                    mb_data, have_more_data, num_elements = self.__model.create_minibatch(data_iter,
+                                                                                          max_num_items=self.__minibatch_size)
+                    if num_elements == 0:

```

Comment:
Thanks! It's at least confusing you're right. I've now tried to make things somewhat simpler. Let me know if these make sense or if you see any way to improve this.

The principle I am using here is that the `data_iter` is gradually consumed. When it's entirely consumed then we stop iterating. During training, we skip the last minibatch that will be smaller than the other ones (because it has fewer samples, the gradient will be more noisy). During validation, we still want to account for partial (non-empty) minibatches to compute the validation score.

Reply:
Yes, this is much clear. Thanks.